contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -291,54 +371,69 @@ fn apply_transfer(
     source_chain: ChainNameRaw,
     destination_chain: ChainNameRaw,
     transfer: &InterchainTransfer,
-) -> Result<(), Error> {
-    subtract_amount_from_source(storage, source_chain, transfer)?;
-    add_amount_to_destination(storage, destination_chain, transfer)
+) -> Result<InterchainTransfer, Error> {
+    let destination_amount = destination_amount(
+        storage,
+        &source_chain,
+        &destination_chain,
+        transfer.token_id,
+        transfer.amount,
+    )?;
+
+    subtract_amount_from_source(storage, source_chain, transfer.token_id, transfer.amount)?;
+    add_amount_to_destination(
+        storage,
+        destination_chain,
+        transfer.token_id,
+        destination_amount,
+    )?;
+
+    Ok(InterchainTransfer {
+        amount: destination_amount,
+
+        ..transfer.clone()
+    })
 }
 
 fn subtract_amount_from_source(
     storage: &mut dyn Storage,
     source_chain: ChainNameRaw,
-    transfer: &InterchainTransfer,
+    token_id: TokenId,
+    amount: nonempty::Uint256,
 ) -> Result<(), Error> {
-    let mut source_instance =
-        try_load_token_instance(storage, source_chain.clone(), transfer.token_id)?;
+    let mut source_instance = try_load_token_instance(storage, source_chain.clone(), token_id)?;
 
     source_instance.supply = source_instance
         .supply
-        .checked_sub(transfer.amount)
+        .checked_sub(amount)
         .change_context_lazy(|| Error::TokenSupplyInvariantViolated {
-            token_id: transfer.token_id,
+            token_id,
             chain: source_chain.clone(),
         })?;
 
-    state::save_token_instance(storage, source_chain, transfer.token_id, &source_instance)
+    state::save_token_instance(storage, source_chain, token_id, &source_instance)
         .change_context(Error::State)
 }
 
 fn add_amount_to_destination(
     storage: &mut dyn Storage,
     destination_chain: ChainNameRaw,
-    transfer: &InterchainTransfer,
+    token_id: TokenId,
+    amount: nonempty::Uint256,
```

Comment:
Think it was cleaner before. Why not just transform the `InterchainTransfer` object before passing into this function?

Reply:
Don't like having function taking unnecessary info.